nixos/modules/services/networking/firewall-nftables.nix

```diff
@@ -0,0 +1,167 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.networking.firewall;
+
+  ifaceSet = concatStringsSep ", " (
+    map (x: ''"${x}"'') cfg.trustedInterfaces
+  );
+
+  portsToNftSet = ports: portRanges: concatStringsSep ", " (
+    map (x: toString x) ports
+    ++ map (x: "${toString x.from}-${toString x.to}") portRanges
+  );
+
+in
+
+{
+
+  options = {
+
+    networking.firewall = {
+      extraInputRules = mkOption {
+        type = types.lines;
+        default = "";
+        example = "ip6 saddr { fc00::/7, fe80::/10 } tcp dport 24800 accept";
+        description = lib.mdDoc ''
+          Additional nftables rules to be appended to the input-allow
+          chain.
+
+          This option only works with the nftables based firewall.
+        '';
+      };
+
+      extraForwardRules = mkOption {
+        type = types.lines;
+        default = "";
+        example = "iifname wg0 accept";
+        description = lib.mdDoc ''
+          Additional nftables rules to be appended to the forward-allow
+          chain.
+
+          This option only works with the nftables based firewall.
+        '';
+      };
+    };
+
+  };
+
+  config = mkIf (cfg.enable && config.networking.nftables.enable) {
+
+    assertions = [
+      {
+        assertion = cfg.extraCommands == "";
+        message = "extraCommands is incompatible with the nftables based firewall: ${cfg.extraCommands}";
+      }
+      {
+        assertion = cfg.extraStopCommands == "";
+        message = "extraStopCommands is incompatible with the nftables based firewall: ${cfg.extraStopCommands}";
+      }
+      {
+        assertion = cfg.pingLimit == null || !(hasPrefix "--" cfg.pingLimit);
+        message = "nftables syntax like \"2/second\" should be used in networking.firewall.pingLimit";
+      }
+      {
+        assertion = config.networking.nftables.rulesetFile == null;
+        message = "networking.nftables.rulesetFile conflicts with the firewall";
+      }
+    ];
+
+    networking.nftables.ruleset = ''
+
+      table inet nixos-fw {
+
+        ${optionalString (cfg.checkReversePath != false) ''
+          chain rpfilter {
+            type filter hook prerouting priority mangle + 10; policy drop;
+
+            meta nfproto ipv4 udp sport . udp dport { 67 . 68, 68 . 67 } accept comment "DHCPv4 client/server"
+            fib saddr . mark ${optionalString (cfg.checkReversePath != "loose") ". iif"} oif exists accept
```

Comment:
Are you sure this is correct? If I understand https://wiki.nftables.org/wiki-nftables/index.php/Matching_routing_information correctly, we should replace `exists` with `eq 0` for strict checking. I don't fully understand what adding `iif` to the lookup does though.

Reply:
```suggestion
            fib saddr . mark ${optionalString (cfg.checkReversePath == "strict") ". iif"} oif ${if cfg.checkReversePath == "strict" then "eq 0" else "exists"} accept
```

Reply:
`exists` is a boolean type in nftables (https://man.archlinux.org/man/extra/nftables/nft.8.en#BOOLEAN_TYPE). Here it’s used to check if the reverse route exists (same as `!= 0`).

Adding `iif` to the lookup will check if the interface that the packet is received on is on the reverse route, which is defined as strict reverse path forwarding in rfc3704.

`iptables-translate` can translate some iptables expression to nftables. e.g.
```
iptables-translate -t mangle -A nixos-fw-rpfilter -m rpfilter --validmark -j RETURN
nft add rule ip mangle nixos-fw-rpfilter fib saddr . mark . iif oif != 0 counter return
```

Reply:
OK, thanks for the explanations!